pubsub/stanpubsub/stan.go

```diff
@@ -0,0 +1,538 @@
+// See https://godoc.org/gocloud.dev/pubsub#hdr-At_most_once_and_At_least_once_Delivery
+// for more background.
+//
+// As
+//
+// stanpubsub exposes the following types for As:
+//  - Topic: stan.Conn
+//  - Subscription: stan.Subscription
+//  - Message.BeforeSend: None.
+//  - Message.AfterSend: None.
+//  - Message: *stan.Msg
+
+package stanpubsub
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/stan.go"
+
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/batcher"
+	"gocloud.dev/pubsub/driver"
+)
+
+func init() {
+	dd = new(defaultDialer)
+	pubsub.DefaultURLMux().RegisterTopic(Scheme, dd)
+	pubsub.DefaultURLMux().RegisterSubscription(Scheme, dd)
+}
+
+// This defaultDialer variable is used only for the test purpose.
+var dd *defaultDialer
+
+// errNotInitialized is an error returned when the topic or subscription is not initialized.
+var errNotInitialized = errors.New("stanpubsub: not initialized")
+
+// defaultDialer dials a default NATS Streaming server based on three required environment variables:
+//	- STAN_SERVER_URL - defines NATS Server Url
+//	- STAN_CLUSTER_ID - defines NATS Streaming Cluster ID
+//	- STAN_CLIENT_ID  - defined NATS Streaming Client ID
+type defaultDialer struct {
+	init   sync.Once
+	opener *URLOpener
+	err    error
+}
+
+// OpenTopicURL implements pubsub.TopicURLOpener interface.
+func (o *defaultDialer) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic, error) {
+	opener, err := o.defaultConn()
+	if err != nil {
+		return nil, fmt.Errorf("open topic %v: failed to open default connection: %v", u, err)
+	}
+	return opener.OpenTopicURL(ctx, u)
+}
+
+// OpenSubscriptionURL implements pubsub.SubscriptionURLOpener interface.
+func (o *defaultDialer) OpenSubscriptionURL(ctx context.Context, u *url.URL) (*pubsub.Subscription, error) {
+	opener, err := o.defaultConn()
+	if err != nil {
+		return nil, fmt.Errorf("open subscription %v: failed to open default connection: %v", u, err)
+	}
+	return opener.OpenSubscriptionURL(ctx, u)
+}
+
+func (o *defaultDialer) defaultConn() (*URLOpener, error) {
+	o.init.Do(o.createdOpener)
+	return o.opener, o.err
+}
+
+func (o *defaultDialer) createdOpener() {
+	serverURL := os.Getenv("STAN_SERVER_URL")
+	if serverURL == "" {
+		o.err = errors.New("STAN_SERVER_URL environment variable not set")
+		return
+	}
+	clusterID := os.Getenv("STAN_CLUSTER_ID")
+	if clusterID == "" {
+		o.err = errors.New("STAN_CLUSTER_ID environment variable not set")
+	}
+	clientID := os.Getenv("STAN_CLIENT_ID")
+	if clientID == "" {
+		o.err = errors.New("STAN_CLIENT_ID environment variable not set")
+	}
+	o.connect(serverURL, clusterID, clientID)
+}
+
+func (o *defaultDialer) connect(serverURL, clusterID, clientID string) {
+	conn, err := stan.Connect(clusterID, clientID, stan.NatsURL(serverURL))
```

Comment:
Does this ever need to reconnect, or does it handle that internally?

Reply:
This is handled internally.